File: tests/fake_xray.py
#!/usr/bin/env python3
import json, socket, struct, sys, threading
from pathlib import Path

args=sys.argv[1:]
if 'version' in args:
    print('Xray 26.7.28 (L-UI test fake)')
    raise SystemExit(0)
try:
    cfg_path=Path(args[args.index('-config')+1])
except Exception:
    print('missing -config')
    raise SystemExit(2)
cfg=json.loads(cfg_path.read_text())
if '-test' in args:
    print('Configuration OK.')
    raise SystemExit(0)
ib=cfg['inbounds'][0]
port=int(ib['port'])
access=Path(cfg.get('log',{}).get('access','/tmp/fake-xray-access.log'))
outbound=(cfg.get('outbounds') or [{'tag':'direct'}])[0].get('tag','direct')
tag=ib.get('tag','test')

ls=socket.socket();ls.setsockopt(socket.SOL_SOCKET,socket.SO_REUSEADDR,1);ls.bind(('127.0.0.1',port));ls.listen(20)

def relay(a,b):
    try:
        while True:
            d=a.recv(65536)
            if not d: break
            b.sendall(d)
    except Exception: pass
    finally:
        try:b.shutdown(socket.SHUT_WR)
        except Exception:pass

def handle(c):
    try:
        h=c.recv(2)
        if len(h)<2:return
        n=h[1]; c.recv(n); c.sendall(b'\x05\x00')
        h=c.recv(4)
        if len(h)<4:return
        atyp=h[3]
        if atyp==1: host=socket.inet_ntoa(c.recv(4))
        elif atyp==3:
            l=c.recv(1)[0]; host=c.recv(l).decode()
        elif atyp==4: host=socket.inet_ntop(socket.AF_INET6,c.recv(16))
        else:return
        dest_port=struct.unpack('!H',c.recv(2))[0]
        r=socket.create_connection((host,dest_port),timeout=5)
        c.sendall(b'\x05\x00\x00\x01\x00\x00\x00\x00\x00\x00')
        access.parent.mkdir(parents=True,exist_ok=True)
        with access.open('a') as f:f.write(f'from tcp:127.0.0.1 accepted tcp:{host}:{dest_port} [{tag} -> {outbound}]\n')
        t=threading.Thread(target=relay,args=(c,r),daemon=True);t.start();relay(r,c)
    finally:
        try:c.close()
        except Exception:pass

while True:
    c,_=ls.accept();threading.Thread(target=handle,args=(c,),daemon=True).start()

File: tests/fake_xray.go
package main

import (
	"encoding/binary"
	"encoding/json"
	"fmt"
	"io"
	"net"
	"os"
	"os/signal"
	"strconv"
	"syscall"
)

type cfg struct {
	Log struct {
		Access string `json:"access"`
	} `json:"log"`
	Inbounds []struct {
		Listen string `json:"listen"`
		Port   int    `json:"port"`
		Tag    string `json:"tag"`
	} `json:"inbounds"`
}

func main() {
	args := os.Args[1:]
	for _, a := range args {
		if a == "version" {
			fmt.Println("Xray 99.99.99 fake")
			return
		}
	}
	configPath := ""
	test := false
	for i, a := range args {
		if a == "-test" {
			test = true
		}
		if a == "-config" && i+1 < len(args) {
			configPath = args[i+1]
		}
	}
	if configPath == "" {
		os.Exit(2)
	}
	data, e := os.ReadFile(configPath)
	if e != nil {
		panic(e)
	}
	var c cfg
	if e = json.Unmarshal(data, &c); e != nil {
		panic(e)
	}
	if test {
		return
	}
	if len(c.Inbounds) == 0 {
		os.Exit(2)
	}
	ib := c.Inbounds[0]
	addr := net.JoinHostPort(ib.Listen, strconv.Itoa(ib.Port))
	ln, e := net.Listen("tcp", addr)
	if e != nil {
		panic(e)
	}
	defer ln.Close()
	sig := make(chan os.Signal, 1)
	signal.Notify(sig, syscall.SIGTERM, syscall.SIGINT)
	go func() { <-sig; ln.Close() }()
	for {
		conn, e := ln.Accept()
		if e != nil {
			return
		}
		go handle(conn, ib.Tag, c.Log.Access)
	}
}
func handle(c net.Conn, inTag, logPath string) {
	defer c.Close()
	head := make([]byte, 2)
	if _, e := io.ReadFull(c, head); e != nil {
		return
	}
	methods := make([]byte, int(head[1]))
	if _, e := io.ReadFull(c, methods); e != nil {
		return
	}
	_, _ = c.Write([]byte{5, 0})
	req := make([]byte, 4)
	if _, e := io.ReadFull(c, req); e != nil {
		return
	}
	host := ""
	switch req[3] {
	case 1:
		x := make([]byte, 4)
		if _, e := io.ReadFull(c, x); e != nil {
			return
		}
		host = net.IP(x).String()
	case 4:
		x := make([]byte, 16)
		if _, e := io.ReadFull(c, x); e != nil {
			return
		}
		host = net.IP(x).String()
	case 3:
		n := make([]byte, 1)
		if _, e := io.ReadFull(c, n); e != nil {
			return
		}
		x := make([]byte, int(n[0]))
		if _, e := io.ReadFull(c, x); e != nil {
			return
		}
		host = string(x)
	default:
		return
	}
	pb := make([]byte, 2)
	if _, e := io.ReadFull(c, pb); e != nil {
		return
	}
	port := int(binary.BigEndian.Uint16(pb))
	up, e := net.Dial("tcp", net.JoinHostPort(host, strconv.Itoa(port)))
	if e != nil {
		_, _ = c.Write([]byte{5, 5, 0, 1, 0, 0, 0, 0, 0, 0})
		return
	}
	defer up.Close()
	_, _ = c.Write([]byte{5, 0, 0, 1, 0, 0, 0, 0, 0, 0})
	if logPath != "" {
		if f, e := os.OpenFile(logPath, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0644); e == nil {
			fmt.Fprintf(f, "tcp:%s accepted [%s -> direct]\n", net.JoinHostPort(host, strconv.Itoa(port)), inTag)
			f.Close()
		}
	}
	done := make(chan struct{}, 1)
	go func() { _, _ = io.Copy(up, c); done <- struct{}{} }()
	_, _ = io.Copy(c, up)
	<-done
}
